### PR TITLE
Support for OAuth

### PR DESCRIPTION
**Pull request by vsevel**
Fixes https://github.com/Apicurio/apicurio-registry-content-sync-operator/issues/140

### CODE REVIEW

**Comment by davgordo** [comment]
I built a test scenario for this PR. Source is here:
https://github.com/davgordo/kafka-quarkus-reference/tree/registry-sync-oauth

Summary:
- sets up a Kafka broker to support Apicurio Registry persistence and application test scenario
- sets up a Keycloak instance, realm, and clients to support the test scenario
- sets up an Apicurio Registry instance with OAuth credentials required for writes
- sets up an s2i build and deployment of @vsevel's feature branch of the content-sync-operator 
- provides an example scenario where an Avro artifact custom resource instance is applied, and a Quarkus application then leverages that artifact

**Comment by vsevel** [comment]
I see the build failed because of:
```
Caused by: com.fasterxml.jackson.databind.exc.InvalidFormatException: Cannot deserialize value of type `java.util.Date` from String "2023-03-23T16:44:55Z": expected format "yyyy-MM-dd'T'HH:mm:ssZ"
 at [Source: (jdk.internal.net.http.ResponseSubscribers$HttpResponseInputStream); line: 1, column: 198] (through reference chain: io.apicurio.registry.rest.v2.beans.ArtifactSearchResults["artifacts"]->java.util.ArrayList[0]->io.apicurio.registry.rest.v2.beans.SearchedArtifact["createdOn"])
```

I am surprised this would have been caused by the changes inside the PR.

**Comment by vsevel** [comment]
any news on this?

**Comment by vsevel** [comment]
hello, any news on this?

**Comment by vsevel** [comment]
> Could you kindly share how you created your keycloak client...? confidential client...?

@davgordo did this part. anything you can share with us David?

**Comment by LittleWat** [comment]
@vsevel Thank you for your reply!

Following the comment, 
https://github.com/davgordo/kafka-quarkus-reference/blob/30edb6d704aa8037124594e35d3e28ae679424d6/kafka/ApicurioRegistry_registry.yaml#L8-L12

I set up the keycloak client (`registry-api`) as a confidential client, and I can confirm that this has worked! ([needed to update the client library, though...](https://github.com/Apicurio/apicurio-registry-content-sync-operator/pull/135))

In [the article](https://www.apicur.io/blog/2021/05/28/registry-security#:~:text=For%20the%20registry%2Dapi%20client%2C%20make%20sure%20to%20configure%20it%20as%20bearer%2Donly), it says
>  For the registry-api client, make sure to configure it as bearer-only. 

But if we want to use the client-credential flows, we must set it as `confidential`  not `bearer-only`. I think this is a little bit confusig point.

**Comment by vsevel** [comment]
I am not familiar with some details about using keycloak 
what I did was testing successfully against an oauth enabled registry, our ADFS and my PR.

**Comment by LittleWat** [comment]
@vsevel I see, cool! I could confirm that this works well with Keycloak!

@andreaTP Sorry to mention you, but could you kindly review this PR, please? 🙏 
This implementation is the same as follows and it would be okay to merge...?
https://github.com/Apicurio/apicurio-registry-examples/blob/9556e7a3bdc8abe5c5a8727de260afbf72fa0b11/rest-client/src/main/java/io/apicurio/registry/examples/SimpleRegistryDemo.java#L47-L56

**Comment by LittleWat** [comment]
@vsevel Sorry, could you rebase to the latest main branch to solve the following CI error?
https://github.com/Apicurio/apicurio-registry-content-sync-operator/actions/runs/4438005146/jobs/7925677302?pr=142#step:9:87
[The PR](https://github.com/Apicurio/apicurio-registry-content-sync-operator/pull/135) is merged, so this error would be solved.

File: sync/src/main/java/io/apicurio/sync/RegistryClientProducer.java [hunk 1]
```diff
@@ -23,7 +30,15 @@ public class RegistryClientProducer {
 
     @Produces
     public RegistryClient registryClient() {
-        return RegistryClientFactory.create(registryUrl);
+        final String tokenEndpoint = System.getenv("AUTH_TOKEN_ENDPOINT");
+        if (tokenEndpoint != null) {
+            final String authClient = System.getenv("AUTH_CLIENT_ID");
+            final String authSecret = System.getenv("AUTH_CLIENT_SECRET");
+            Auth auth = new OidcAuth(new JdkHttpClient(tokenEndpoint, Collections.emptyMap(), null, new AuthErrorHandler()), authClient, authSecret);
+            return RegistryClientFactory.create(new JdkHttpClient(registryUrl, Collections.emptyMap(), auth, new AuthErrorHandler()));
```
Comment (LittleWat) [inline sync/src/main/java/io/apicurio/sync/RegistryClientProducer.java:38]: After deep diving,  I found that this handler should be `AuthErrorHandler` to catch `ArtifactAlreadyExistsException` (e.g. https://github.com/Apicurio/apicurio-registry-content-sync-operator/blob/579db7e391088a3084367d14c2e4e14de344ceb0/sync/src/main/java/io/apicurio/sync/controller/ArtifactController.java#L260)


```suggestion
            Auth auth = new OidcAuth(new JdkHttpClient(tokenEndpoint, Collections.emptyMap(), null, new io.apicurio.registry.rest.client.impl.ErrorHandler()), authClient, authSecret);
            return RegistryClientFactory.create(new JdkHttpClient(registryUrl, Collections.emptyMap(), auth, new io.apicurio.registry.rest.client.impl.ErrorHandler()));
```

I created a PR for building the image that supports OAuth and confirmed that it worked as expected.
https://github.com/Apicurio/apicurio-registry-content-sync-operator/pull/144
(Though, it has some issues in testing and generating beans.)

**Comment by vsevel** [comment]
anything needed from me?

**Comment by vsevel** [comment]
any news on this?

**Comment by LittleWat** [comment]
The dependency problem is too complicated 😢  ... I prefer [the Golang version](https://github.com/dweber019/apicurio-registry-artifact-operator)...

BTW, I found it a little bit challenging to solve [the CI error](https://github.com/Apicurio/apicurio-registry-content-sync-operator/actions/runs/4841984776/jobs/8628631936?pr=144#step:5:479) by myself 🤦 
So I created an issue for this. Any help is really appreciated 🙏 
- https://github.com/Apicurio/apicurio-common-rest-client/issues/145

**Comment by carlesarnal** [comment]
I have added https://github.com/Apicurio/apicurio-registry-content-sync-operator/pull/145 for adding oauth support. Please keep in mind that most dependencies are still outdated, but CI passes for this support. As a result, I'm closing this PR. @LittleWat @vsevel if you want to make new contributions to the repository it should be a bit easier now. Thanks!

**Pull request closed without merging**